### PR TITLE
Experiment CRD v1alpha2

### PR DESCRIPTION
**Pull request by kimikowang**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -35,6 +36,10 @@ deploy: manifests
 	  --name iter8-controller \
 	  --set image.repository=`echo ${IMG} | cut -f1 -d':'` \
 	  --set image.tag=`echo ${IMG} | cut -f2 -d':'` \
+	  -x templates/default/namespace.yaml \
+	  -x templates/crds/${CRD_VERSION}/iter8.tools_experiments.yaml \
+	  -x templates/metrics/iter8_metrics.yaml \
+	  -x templates/notifier/iter8_notifiers.yaml\
```
Comment (kalantar) [inline Makefile:42]: Why is this addition necessary? Isn't it covered by the dependency on `manifests` (cf. line 34)

File: pkg/apis/iter8/v1alpha2/experiment_types.go [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+	Criteria []Criterion `json:"criteria,omitempty"`
+
+	// Endpoint of reaching analytics service
+	// default is http://iter8-analytics.iter8
+	// +optional
+	AnalyticsEndpoint string `json:"analyticsEndpoint,omitempty"`
+
+	// Duration specifies how often/many times the expriment should re-evaluate the assessment
+	// +optional
+	Duration `json:"duration,omitempty"`
```
Comment (kalantar) [inline pkg/apis/iter8/v1alpha2/experiment_types.go:72]: `Duration` is missing a type.

File: pkg/apis/iter8/v1alpha2/experiment_types.go [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+	Metric string `json:"metric"`
+
+	// Threshold specifies the numerical value for a success criterion
+	// Metric value above threhsold violates the criterion
+	// +optional
+	Threshold Threshold `json:"threshold,omitempty"`
+
+	// Once a target violates this criterion, traffic to it will be cutoff or not
+	// +optional
+	CutoffTrafficOnViolation bool `json:"cutoffTrafficOnViolation,omitempty"`
```
Comment (kalantar) [inline pkg/apis/iter8/v1alpha2/experiment_types.go:107]: This should be part of a `Threshold`. See https://github.com/iter8-tools/iter8-controller/issues/97#issuecomment-616701414

File: pkg/apis/iter8/v1alpha2/experiment_types.go [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+// Duration specifies how often/many times the expriment should re-evaluate the assessment
+type Duration struct {
+	// Interval specifies duration between iterations
+	// default is 30s
+	// +optional
+	Interval string `json:"interval,omitempty"`
+	// Iterations indicates the amount of iteration
+	// default is 100
+	// +optional
+	Iterations int32 `json:"iterations,omitempty"`
```
Comment (kalantar) [inline pkg/apis/iter8/v1alpha2/experiment_types.go:135]: This should be `maxIterations`, not `iterations`. See https://github.com/iter8-tools/iter8-controller/issues/97#issuecomment-616701414

File: pkg/apis/iter8/v1alpha2/experiment_types.go [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+
+	// OnTermination determines traffic split status at the end of experiment
+	// +kubebuilder:validation:Enum={to_winner,to_baseline,keep_last}
+	// +optional
+	OnTermination string `json:"onTermination,omitempty"`
+
+	// Only requests fulfill the match section would be used in experiment
+	// Istio matching rules are used
+	// +optional
+	Match `json:"match,omitempty"`
```
Comment (kalantar) [inline pkg/apis/iter8/v1alpha2/experiment_types.go:154]: `Match` is missing a type.

File: pkg/apis/iter8/v1alpha2/experiment_types.go [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+	Spec ExperimentSpec `json:"spec,omitempty"`
+
+	// +optional
+	Status ExperimentStatus `json:"status,omitempty"`
+
+	// +optional
+	Metrics ExperimentMetrics `json:"metrics,omitempty"`
+
+	// +optional
+	ManualOverride `json:"manualOverride,omitempty"`
```
Comment (kalantar) [inline pkg/apis/iter8/v1alpha2/experiment_types.go:43]: `ManualOverride` is missing a type.

File: pkg/apis/iter8/v1alpha2/experiment_types.go [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+	// default is 100
+	// +optional
+	Iterations int32 `json:"iterations,omitempty"`
+}
+
+// TrafficControl specifies constrains on traffic and stratgy used to update the traffic
+type TrafficControl struct {
+	// Strategy used to shift traffic
+	// default is progressive_traffic_shift
+	// +kubebuilder:validation:Enum={progressive_traffic_shift, top_2, uniform}
```
Comment (kalantar) [inline pkg/apis/iter8/v1alpha2/experiment_types.go:142]: Propose to change `progressive_traffic_shift` to `progressive`. See https://ibm-research.slack.com/archives/GA0KVAGQJ/p1588794125037800

File: pkg/apis/iter8/v1alpha2/experiment_types.go [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+
+	// Cleanup indicates whether routing rules and deployment receiving no traffic should be deleted at the end of experiment
+	// +optional
+	Cleanup bool `json:"cleanup"`
+}
+
+// Service is a reference to the service that this experiment is targeting at
+type Service struct {
+	// defines the characteristics of the service
+	*corev1.ObjectReference `json:",inline"`
```
[the remark this answers is not in the document]
Comment (kimikowang) [reply pkg/apis/iter8/v1alpha2/experiment_types.go:86]: `apiVersion` is optional..

File: Makefile (hunk 1 above)
Comment (kimikowang) [reply Makefile:42]: manifests just generates the file. crd of different versions stored in different directories.

File: pkg/apis/iter8/v1alpha2/experiment_types.go (hunk 2 above)
Comment (kimikowang) [reply pkg/apis/iter8/v1alpha2/experiment_types.go:72]: This is a type. Golang allows auto field name from type.

File: pkg/apis/iter8/v1alpha2/experiment_types.go [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+	// Strategy used to shift traffic
+	// default is progressive_traffic_shift
+	// +kubebuilder:validation:Enum={progressive_traffic_shift, top_2, uniform}
+	// +optional
+	Strategy string `json:"strategy,omitempty"`
+
+	// OnTermination determines traffic split status at the end of experiment
+	// +kubebuilder:validation:Enum={to_winner,to_baseline,keep_last}
+	// +optional
+	OnTermination string `json:"onTermination,omitempty"`
```
Comment (kalantar) [inline pkg/apis/iter8/v1alpha2/experiment_types.go:149]: Valid value `keep_last` should be `keep_last_traffic_split`. See https://ibm-research.slack.com/archives/GA0KVAGQJ/p1588794267040200

File: pkg/apis/iter8/v1alpha2/experiment_types.go (hunk 7 above)
Comment (kalantar) [reply pkg/apis/iter8/v1alpha2/experiment_types.go:142]: That conversation includes additional suggestions to valid values. It suggests `promote_winner` and `promote_baseline` as well.

File: pkg/apis/iter8/v1alpha2/experiment_types.go [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+	// +optional
+	CurrentIteration int `json:"currentIteration,omitempty"`
+
+	// GrafanaURL is the url to the Grafana Dashboard
+	// +optional
+	GrafanaURL string `json:"grafanaURL,omitempty"`
+
+	// Assessment returned by the last analyis
+	// +optional
+	Assessment `json:"assessment,omitempty"`
```
Comment (kalantar) [inline pkg/apis/iter8/v1alpha2/experiment_types.go:223]: `Assessment` is missing a type

File: pkg/apis/iter8/v1alpha2/experiment_types.go [hunk 11]
```diff
@@ -0,0 +1,327 @@
...
+
+	// TrafficSplit tells the current traffic spliting among targets
+	// +optional
+	TrafficSplit `json:"trafficSplit,omitempty"`
```
Comment (kalantar) [inline pkg/apis/iter8/v1alpha2/experiment_types.go:227]: `TrafficSplit` is missing a type

File: pkg/apis/iter8/v1alpha2/experiment_types.go [hunk 12]
```diff
@@ -0,0 +1,327 @@
...
+
+	// Phase marks the Phase the experiment is at
+	// +optional
+	Phase `json:"phase,omitempty"`
```
Comment (kalantar) [inline pkg/apis/iter8/v1alpha2/experiment_types.go:231]: Phase is missing a type

File: pkg/apis/iter8/v1alpha2/experiment_types.go [hunk 13]
```diff
@@ -0,0 +1,327 @@
...
+
+	// Percentage specifies the amount of traffic to service that would be used in experiment
+	// default is 100
+	// +optional
+	Percentage int32 `json:"percentage,omitempty"`
+
+	// MaxIncrement is the upperlimit of traffic increment for a target in one iteration
+	// default is 2
+	// +optional
+	MaxIncrement int32 `json:"maxIncrement,omitempty"`
```
Comment (kalantar) [inline pkg/apis/iter8/v1alpha2/experiment_types.go:164]: Discussion https://ibm-research.slack.com/archives/GA0KVAGQJ/p1588795333048700 suggests these should be `numeric` allowing user to specify either as float or int. But only int are acceptable. Not sure what is best. @sriumcp @kimikowang

**Comment by kalantar** [comment]
@kimikowang why was this closed?
